contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+                      |                                                   set result to failure
+                      |            validate <---------------------------- PUT /pods
+                      |            admission(init):
+                      |              check authz
+                      v            signal failure
+                  observe <------- save etcd
+response <------- handle               |
+                                       v
+                                   delete object
+```
+
+
+##### Quota
+
+Quota consists of two distinct parts - object count, which prevents abuse for some limited
+resource, and sub-object consumption which may not be known until after all initializers
+are executed.
+
+Object count quota should be applied prior to the object being persisted to etcd (prior
+to initialization). All other quota should be applied when the object completes initialization.
+Compensation should run at the normal spot.
+
+
+##### Bypassing initializers
+
+Initializers, like external admission hooks, raise the potential for a cluster that cannot
+make progress or heal itself. An initializer on pods could block an emergency fallback
+scheduler from launching a new scheduler pod. An initializer on the endpoints resource could
+prevent masters from registering themselves, blocking an extension API server from observing
+endpoints changes to allow it to watch endpoints. In general, new initializers must be
+careful to not create circular dependencies with the masters.
+
+The ability to set an empty initializer list allows cluster level components to make progress
+in the face of extension. Additional information may need to be returned with an object
+creation error to indicate which component failed to initialize.
+
+
+#### Generic external admission webhook
+
+Existing webhooks demonstrate specific admission callouts for image policy. In general, the `admission.Interface`
+already defines a reasonable pattern for mutations. Admission is much less common for retrieval operations.
+
+Add a new `GenericAdmissionWebhook` that is a list of endpoints that will receive POST operations. The schema
+of the object is modelled after `SubjectAccessReview` and `admission.Interface`. The caller posts the object
+to the server and expects a 200, 204 or 400 response. If the response is 200 or 204, the caller proceeds. If
+the response is 400, the client should interpret the response body to determine the "status" sub field and
+then use that as the response.
+
+```
+POST [arbitrary_url]
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "spec": {
+        "resource": "pods",
+        "subresource": "",
+        "operation": "Create",
+        "object": {...},
+        "oldObject": nil,
+        "userInfo": {
+            "name": "user1",
+            "uid": "cn=user,dn=ou",
+            "groups": ["g=1"]
+        }
+    },
+    "status": {
```

Comment:
I was just calling it out here because I didn't create a schema in the doc yet - it would not be required on POST, but would be in the POST response.

Note as well that nothing in this assumes that external admission must always use a webhook style call.  A GRPC backend is totally reasonable and can be added at any time without invalidating this proposal.